### PR TITLE
feat: Make IgnoreFile stateful to allow for overrides

### PR DESCRIPTION
**Pull request by abelmokadem**
I faced an issue that did not allow me to override the gitignore file so that it would ignore the projen generated files. This is due to how the IgnoreFile currently works. With the modification I made, we can allow users to override the IgnoreFile anywhere down the road. 

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by Chriscbr** [comment]
Thanks for writing up this contribution. :-) I think it would be nice to also have some options for exploring

File: src/ignore-file.ts [hunk 1]
```diff
@@ -15,24 +19,21 @@ export class IgnoreFile extends FileBase {
    * @param patterns Patterns to exclude from git commits.
    */
   public exclude(...patterns: string[]) {
-    this.excludes.push(...patterns);
+    patterns.forEach(pattern => this.patterns[pattern.replace(/^\!/, '')] = IgnorePatternType.EXCLUDE)
```
Comment (Chriscbr) [inline src/ignore-file.ts:22]: What is the purpose of removing the exclamation marks here?

File: src/ignore-file.ts [hunk 2]
```diff
@@ -1,10 +1,14 @@
 import { FileBase, IResolver } from './file';
 import { Project } from './project';
-import { dedupArray } from './util';
+
+export enum IgnorePatternType {
+    INCLUDE,
+    EXCLUDE
+}
```
Comment (Chriscbr) [inline src/ignore-file.ts:7]: Since these are technically only used internally, is it possible we could leave this private?

**Review by Chriscbr — COMMENTED** [review]
Could we also add a unit test to show what happens if you exclude a pattern and then include it (or vice versa)?

File: src/ignore-file.ts (hunk 1 above)
Comment (abelmokadem) [reply src/ignore-file.ts:22]: Good question. I can imagine that someone would think that they have to use the question mark in order to include a file. I could leave the logic out, but then it's up to the user to not make the mistake.

Without removing the question mark you get the following result:
```
exclude('/my-file.js') // this.patterns = {'/my-file.js': 'EXCLUDE'}
include('!/my-file.js') // this.patterns = {'/my-file.js': 'EXCLUDE', '!/my-file.js': 'INCLUDE'}
```

If you remove the question mark you get:
```
exclude('/my-file.js') // this.patterns = {'/my-file.js': 'EXCLUDE'}
include('!/my-file.js') // this.patterns = {'/my-file.js': 'INCLUDE'}
```

File: src/ignore-file.ts (hunk 2 above)
Comment (abelmokadem) [reply src/ignore-file.ts:7]: I can make this private by not exporting the enum right?

**Comment by abelmokadem** [comment]
@Chriscbr I have made the enum private and made the test cases more explicit. I also noticed a small flaw in the first test case. The results were already sorted so there was no way to know if the test was failing or not.

**Comment by abelmokadem** [comment]
Test against my own project. I'm now able to ignore files that were generated by projen 😁

**Review by Chriscbr — APPROVED** [review]

**Pull request merged**